### DISCUSSION TITLE
Allow for "direct query" style queries

### DISCUSSION THREAD

**Discussion by creativedutchmen** [evidence-dev/evidence#2333] [Ideas]
## Feature Description

While the choice of pre-loading all the data together with DuckDB makes the experience of using evidence quite snappy, it also means that some types of dashboard are not viable - simply because they would require loading way too much data. It would be great if it were possible to allow some queries to hit the database directly.

While it might seems undesirable, the number of OLAP databases that could handle this type of workload has exploded, and it would make the deployment easier and solution space (much) larger.

## Goal of Feature

Expand the solution space from pre-aggregated data that fits easily into memory to the full set of (ad-hoc) questions one could ask from the data. For instance, drill downs per product, country, membership status, ...

## Current Solution / Workarounds

Currently we need to have a ton of queries that precompute the drill downs we think are interesting, and disallow more than a single variable in these drill downs (ie sales per brand, sales per country, sales per channel)

## Alternatives

Use evidence as a nice looking (very basic) email report, and use another tool for the deep-dives. Super unsatisfying, because it would require multiple tools again (and those tools can usually also send emails)

**Comment by creativedutchmen** [comment]
Apart from the cases where the current flow is simply not feasible, it would also make reports on large tables more editor-friendly.

For instance, imagine that the `orders` table in the example project is billions of rows. You wouldn't be able to do `select * from orders` anymore, and instead would have to use a large rollup table (which might be too big, still) or flat tables per query (`daily_orders_per_country.sql`)

**Comment by mcrascal** [comment]
I like this. At the moment we're focused on building out what you can do in the client side run time. For example, more interactive components like pivot tables and the recent [dimension grid](https://docs.evidence.dev/components/dimension-grid), as well as increasing the interactive performance from supporting source tables with ~1 million records to 10-20 million, but at some point we will likely explore pushing queries down to the data warehouse.

**Comment by kylelundstedt** [comment]
If I'm correctly following the logic described in the Universal SQL article, the browser-based DuckDB WASM currently is limited to querying ONLY the Parquet files created by `npm run source.  Is that correct?

Given that DuckDB WASM has the ability to directly query files in object storage (e.g., partitioned Parquet files on AWS S3) or remote tables in MotherDuck, what would prevent Evidence's DuckDB WASM from using the cached Parquet file(s) for some queries and, say, remote Parquet files on S3 for drill downs to more detail?

I can imagine that making this local/remote Data Source setup would be more complicated for other databases such as Snowflake or BigQuery (e.g., pushing queries down to the data warehouse, per @mcrascal).  However, it seems that the DuckDB and MotherDuck sources might be straightforward, given DuckDB's built-in ability to query files in cloud object storage.  Please let me know if I'm missing something.  😁

Our use case involves Evidence reports built upon Parquet files (soon to be Iceberg formatted) hosted on S3, with each file (table) partitioned by day.  We've built Evidence reports from this S3-hosted data that create cached Parquet files containing daily tables for the current day, as well as aggregated monthly tables for the most recent 6 months.

These cached Parquet files make our reports render quite quickly, and give us the filter interactivity that we need.  However, for example, we'd like to give users the ability to look at daily tables for alternative days, without having to create cached Parquet files for all of those days.  Similarly, our users sometimes will be looking at a report based on a set of monthly tables that are cached, and then want to drill down into data from  one or more daily tables that are not cached.  In these circumstances, adding the interactivity delay of waiting for queries against tables on S3 would be no problem.

So, we'd love to have a DuckDB-based Data Source setup so that it automatically caches the current daily tables to Parquet in the browser, yet pushes any queries to non-cached daily tables off to the partitioned-by-day Parquet files in object storage.  Does that make sense for Evidence?

**Comment by sgarbesi** [comment]
Great tool, but I don't see how it can possibly work at scale in a production system without direct queries.

Maybe I'm wrong and someone can help me here. If we wanted to display a customer usage page where a stakeholder can filter by a date range to see the usage for all customers across the entire product, how could you pre-cache this locally? For example there's a date component dropdown at the top of the page allowing the browser client to filter the date range (not a statically written date range in a SQL query, it's dynamic.) The only way I think that you could do this (with the tool in its current state) is by pulling down the entire customer usage catalog of data to local, which could be 100s of million of rows and GBs of data. Then every time we need the data to be updated we need to re-pull down the entire dataset again? Doesn't seem feasible.

For smaller things, looks really cool. Sadly, can't realistically see using this in production due to this limitation I mentioned.

Hope I got it wrong and it can be used like this!

**Reply by clementlefevre** [reply]
First, evidence is an amazing tool for fast, slick and versatile dashboarding. The dev/deploy speed is just incredible.
Before using it, i had my own solution with a basic golang backend  calling the duckdb cli  and serving a svelte frontend that triggered via REST the sql queries. 
Ideally, having the possiblity to switch from the wasm duckdb/preprocessed parquet to REST calls with sql as payload would be a game changer. Not sure how complex/safe such option would be, but definitely worth when scaling comes in play. 

Hats off to all the contributors here ! 👏 👏 👏

**Comment by bluepuma77** [comment]
I want to add my support for this. I initially thought Evidence was a simple and elegant tool: write SQL and charts together in a single file, get a great-looking report, and move quickly. That part is really compelling.

But after using it more, I realized the current model requires deciding the SQL queries up front and caching the resulting data locally. That works for smaller dashboards or predefined reports, but it breaks down during exploration on huge databases, where I often do not know the exact drilldowns or filters ahead of time. Direct query support, even as an optional mode for specific queries or drilldowns, would make Evidence much more useful for production-scale analytics while keeping the current fast cached workflow where it works well.

Another related discussion: https://github.com/evidence-dev/evidence/discussions/3056
